ArduCopter/mode_guided.cpp

```diff
@@ -768,6 +779,163 @@ bool ModeGuided::limit_check()
+            // failure to set destination can only be because of missing terrain data
+            copter.failsafe_terrain_on_event();
+        }
+
+        // if we are outside the circle, point at the edge, otherwise hold yaw
+        const Vector3f &circle_center_neu = copter.circle_nav->get_center();
+        const Vector3f &curr_pos = inertial_nav.get_position();
+        float dist_to_center = norm(circle_center_neu.x - curr_pos.x, circle_center_neu.y - curr_pos.y);
+        // initialise yaw
+        // To-Do: reset the yaw only when the previous navigation command is not a WP.  this would allow removing the special check for ROI
+        if (auto_yaw.mode() != AUTO_YAW_ROI) {
+            if (dist_to_center > copter.circle_nav->get_radius() && dist_to_center > 500) {
+                auto_yaw.set_mode_to_default(false);
+            } else {
+                // vehicle is within circle so hold yaw to avoid spinning as we move to edge of circle
+                auto_yaw.set_mode(AUTO_YAW_HOLD);
+            }
+        }
+    } else {
+        circle_start();
+    }
+
+}
+
+void ModeGuided::circle_start()
+{
+    guided_mode = Guided_Circle;
+
+    // initialise circle controller
+    copter.circle_nav->init(copter.circle_nav->get_center(), copter.circle_nav->center_is_terrain_alt());
+
+    if (auto_yaw.mode() != AUTO_YAW_ROI) {
+        auto_yaw.set_mode(AUTO_YAW_CIRCLE);
+    }
+}
+
+void ModeGuided::circle_run()
+{
+    // process pilot's yaw input
+    float target_yaw_rate = 0;
+    if (!copter.failsafe.radio && use_pilot_yaw()) {
+        // get pilot's desired yaw rate
+        target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+        if (!is_zero(target_yaw_rate)) {
+            auto_yaw.set_mode(AUTO_YAW_HOLD);
+        }
+    }
+
+    // call circle controller
+    copter.failsafe_terrain_set_status(copter.circle_nav->update());
+
+    // call z-axis position controller
+    pos_control->update_z_controller();
+
+    if (auto_yaw.mode() == AUTO_YAW_HOLD) {
+        // roll & pitch from waypoint controller, yaw rate from pilot
+        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(copter.circle_nav->get_roll(), copter.circle_nav->get_pitch(), target_yaw_rate);
+    } else {
+        // roll, pitch from waypoint controller, yaw heading from auto_heading()
+        attitude_control->input_euler_angle_roll_pitch_yaw(copter.circle_nav->get_roll(), copter.circle_nav->get_pitch(), auto_yaw.yaw(), true);
+    }
+}
+
+void ModeGuided::circle_check(uint8_t count)
+{
+    // check if we've reached the edge
+    if (mode() == Guided_CircleMoveToEdge) {
+        if (copter.wp_nav->reached_wp_destination()) {
+            // start circling
+            circle_start();
+        }
+        return ;
+    }
+    // check if we have completed circling
+    if (fabsf(copter.circle_nav->get_angle_total()/M_2PI) >= count) {
+        if(mode() != Guided_Angle) {
```

Comment:
I am pretty sure we want WP mode here and not Angle mode

Reply:
actually, I had tried using Guided_WP but there was a small issue - after completion of the loiter turns if we set a new wp the drone does not align itself properly to the new path. Although the drone does move in the proper path, I think the yaw is not properly set.
![map4](https://user-images.githubusercontent.com/62841337/106564232-fa89f180-6552-11eb-8cb7-890f4d59e20c.png)
I didn't find the above issue while going into Guided_Angle